### DISCUSSION TITLE
Unsupported dpIdTypes in TuyaMCU

### DISCUSSION THREAD

**Discussion by OgnjenX** [arendst/Tasmota#12089] [Ideas]
On the device I flashed with TuyaMCU, there are 3 channels that are currently not supported by Tasmota (no TuyaSend<x> command can control it properly). One of them has dpIdType5 and the other two are recognized as dpIdType0. I would like to know if there is currently any possibility of controlling them that are not recognized as one of Integer, Enum, String, or Boolean types.

**Comment by ascillato2** [comment]
Hi, Can you post some logs? Can you tell us which device it is?

**Reply by OgnjenX** [reply]
Device is http://www.cn-hysen.com/productinfo/451126.html
Channels that are not controllable from Tasmota (or at least I don't know how) have dpid 13, 108, and 109.
DpIdType of 108 and 109 is 0 (DpIdType = 0).
DpIdType of 13 is 5 (DpIdType = 5).
My focus is on controlling channels 108 and 109.
I'm not sure, but my guess based on this
<img width="904" alt="image" src="https://user-images.githubusercontent.com/48242642/118155159-35d07200-b418-11eb-9d7a-ce9c13eccf8c.png">
is that 0 could be string value, and 5 could be fault data-report only.

Channels with dpId 108 and 109, on this device, are used for sending schedule commands (hours, minutes, and wanted temperature). Channel 108 is used for workdays, 109 for weekends. The format is next (if I understood it correctly from analysis): each of these 3 values (hours, minutes, and temperature) is represented as 2 digit hexadecimal number grouped together in 6 digit hex number - which represents information for the one-day timer. Days are grouped one by one, and the whole message is sent to the channel as a hex number (same for 108 and 109).

Some logs when changes happen on these two channels:

18:34:02.448 TYA: Heartbeat
18:34:12.325 {"TuyaReceived":{"Data":"55AA000700106C00000C0F281814351D14361F14372017","Cmnd":7,"CmndData":"6C00000C0F281814351D14361F143720","DpType0Id108":"0x0F281814351D14361F143720","108":{"DpId":108,"DpIdType":0,"DpIdData":"0F281814351D14361F143720"}}}
18:34:12.329 TYA: fnId=0 is set for dpId=108
18:34:12.349 {"TuyaReceived":{"Data":"55AA000700106D00000C07010F13291B13351E14361ECB","Cmnd":7,"CmndData":"6D00000C07010F13291B13351E14361E","DpType0Id109":"0x07010F13291B13351E14361E","109":{"DpId":109,"DpIdType":0,"DpIdData":"07010F13291B13351E14361E"}}}
18:34:12.353 TYA: fnId=0 is set for dpId=109
18:34:13.421 TYA: Send "55aa00000000ff"
18:34:13.434 {"TuyaReceived":{"Data":"55AA000000010101","Cmnd":0,"CmndData":"01"}}
18:34:13.437 TYA: Heartbeat
18:34:16.412 TYA: Send "55aa001c00080115050d1222100493"
18:34:21.116 {"TuyaReceived":{"Data":"55AA000700106C00000C0F181814351D14361F14372007","Cmnd":7,"CmndData":"6C00000C0F181814351D14361F143720","DpType0Id108":"0x0F181814351D14361F143720","108":{"DpId":108,"DpIdType":0,"DpIdData":"0F181814351D14361F143720"}}}
18:34:21.119 TYA: fnId=0 is set for dpId=108
18:34:21.140 {"TuyaReceived":{"Data":"55AA000700106D00000C07010F13291B13351E14361ECB","Cmnd":7,"CmndData":"6D00000C07010F13291B13351E14361E","DpType0Id109":"0x07010F13291B13351E14361E","109":{"DpId":109,"DpIdType":0,"DpIdData":"07010F13291B13351E14361E"}}}
18:34:21.144 TYA: fnId=0 is set for dpId=109
18:34:21.540 WIF: Checking connection...


If I try to control them with the TuyaSend command (best guess would be TuyaSend3 - because they are most likely strings, but I also tried all other possibilities, to be sure), I get values that are not valid - out of range or something similar, and if I change the value that is being sent, result changes, but not reasonable (for example if I change first two digits, hours, minutes and temperature is changed, and in the most cases, they also get invalid value.

**Comment by OgnjenX** [comment]
Any suggestions on this topic, please? :)

**Reply by barbudor** [reply]
What are the logs when use TuyaSend3?
From your analysis what should be the complete message that should be sent (from 0055 to ending CRC byte)? 
Can you try using SerialSend5 with such a complete string?

**Reply by OgnjenX** [reply]
When I try to send commands with TuyaSend3 (I copied the message from the console when I changed the schedule manually from device display - 0x062D15092D140C0017130018 , and changed the first occasion of 6 with 8 - so expected result would be that timer has been moved from 6 to 8 am) I get success in the console:

18:50:01.679 CMD: Group 0, Index 3, Command "TUYASEND", Data "108,"0x082D15092D140C0017130018""
18:50:01.725 TYA: Send "55aa000600206c03001c223078303832443135303932443134304330303137313330303138228c"
18:50:01.729 RSL: RESULT = {"TuyaSend":"Done"}
18:50:01.757 {"TuyaReceived":{"Data":"55AA000700106C00000C223023303823443123303923B2","Cmnd":7,"CmndData":"6C00000C223023303823443123303923","DpType0Id108":"0x223023303823443123303923","108":{"DpId":108,"DpIdType":0,"DpIdData":"223023303823443123303923"}}}

But as you can see, the result makes no sense (timers are not just set to wrong value, but to invalid value on display - e.g. 35:45 , or sometimes totally wrong characters appear).

I tried using TuyaSend5, but got command error in console:

8:50:17.645 CMD: Group 0, Index 5, Command "TUYASEND", Data "108,0x082D15092D140C0017130018"
18:50:17.650 RSL: RESULT = {"Command":"Error"}

**Reply by barbudor** [reply]
In `18:50:01.725 TYA: Send "55aa000600206c03001c223078303832443135303932443134304330303137313330303138228c"`
`22307830383244313530393244313430433030313731333030313822` is actually the ASCII for `"0x082D15092D140C0017130018"`
Looking at the code of `xdrv_16_tuyamcu.ino` it appears that `TuyaSend3` is sending the Text provided, not converting it from HEX to binary

That's why I suggested you to forge manually the whole Tuya string from header to CRC and use `SerialSend5` and not `TuyaSend5`
Something like : `SerialSend5 55aa000600206c03000c082D15092D140C0017130018**` where `**` is the checksum that you would have to calculate by hand as the sum of every byte of the payload (from 08 to 18)

It looks like there is no TuyaSend command for sending HEX strings (or I missed it)
At least if you confirm that sending manually forged string works, we could look at adding such command

**Reply by barbudor** [reply]
I've made a test version with a new `TuyaSend5` command to send an hex string
Would work such like : `TuyaSend5 108, 082D15092D140C0017130018` (note there is neither `"` nor `0x`)

```
20:03:13.218 CMD: TuyaSend5 108, 082D15092D140C0017130018
20:03:13.220 SRC: WebConsole from 192.168.168.5
20:03:13.224 CMD: Grp 0, Cmnd 'TUYASEND', Idx 5, Len 29, Data '108, 082D15092D140C0017130018'
20:03:13.252 TYA: Send "55aa000600106c03000c082d15092d140c001713001872"
20:03:13.257 MQT: stat/nodemcu/TUYASEND = {"TuyaSend":"Done"}
```

Should do what you need.

Are you using a prebuilt tasmota binary ? which one ? `tasmota.bin(.gz)` ?
Or are you building your own ?
I can provide you either a prebuilt binary or the modified tuyamcu.ino driver for testing before doing a PR

**Reply by OgnjenX** [reply]
I'm using prebuilt tasmota.bin (version 9.4.0). It would be great if you can provide a binary or modified file (both are fine). Thank you very much :)

**Reply by sfromis** [reply]
IIRC, sending hex commands have been accomplished in a more cumbersome way, by running as standard serial (SerialSend5) instead of "Tuya mode".

**Comment by barbudor** [comment]
Here you are, tasmota.bin and tasmota.bin.gz in the zip file

[firmware.zip](https://github.com/arendst/Tasmota/files/6556269/firmware.zip)

**Reply by OgnjenX** [reply]
It works! Thank you very much.

**Reply by barbudor** [reply]
Pull request https://github.com/arendst/Tasmota/pull/12211 submitted

Should it be approved, the new command will be part of the development version a few hours later.
